rfcs/0084-input-aware-fetchers.md

```diff
@@ -0,0 +1,147 @@
+```
+/nix/store/l98gjfznp8lpxi0hvj4i0rw34xnnqma8-source
+```
+
+However, if we now intuitively update the revision of the expression to the new 2.3.10 release like
+```nix
+(import <nixpkgs> {}).fetchFromGitHub {
+  owner = "NixOS";
+  repo = "nix";
+  rev = "2.3.10";
+  sha256 = "0d4c3ddpqa1q4j15cl8d7g3igiw6clqczf8dcp4pbpvlm9a64rki";
+}
+```
+
+We can still build this successfully, getting the exact same store path
+```
+/nix/store/l98gjfznp8lpxi0hvj4i0rw34xnnqma8-source
+```
+
+This can't be right! Of course, people with Nix knowledge will tell you that you need to change the hash to something invalid first in order for it to attempt a refetch. This is because Nix already has a path with that exact output `sha256` hash, so it just reuses that. Indeed, changing the output hash to something invalid works:
+```nix
+(import <nixpkgs> {}).fetchFromGitHub {
+  owner = "NixOS";
+  repo = "nix";
+  rev = "2.3.10";
+  sha256 = "0000000000000000000000000000000000000000000000000000";
+}
+```
+
+Building this we get a hash mismatch as expected:
+
+```
+hash mismatch in fixed-output derivation '/nix/store/5d3k20pzgjyccmpqfina1cvbl28zxz6a-source':
+  wanted: sha256:0000000000000000000000000000000000000000000000000000
+  got:    sha256:1cx9yv62rylfv8p09pidsmqy8qim1bbjaa8pj1j8xj7vkrm0dri1
+```
+
+Now we can copy the hash into the expression:
+```nix
+(import <nixpkgs> {}).fetchFromGitHub {
+  owner = "NixOS";
+  repo = "nix";
+  rev = "2.3.10";
+  sha256 = "1cx9yv62rylfv8p09pidsmqy8qim1bbjaa8pj1j8xj7vkrm0dri1";
+}
+```
+
+And build it successfully to get the correct result in the output path
+```
+/nix/store/5d3k20pzgjyccmpqfina1cvbl28zxz6a-source
+```
+
+It is however very easy to forget to update the hash to something invalid, which as many know has been a source of much confusion, questions and frustration.
+
+However it doesn't have to be this way!
+
+# Detailed design
+[design]: #detailed-design
+
+This problem can be solved relatively easily by realizing that Nix not only uses the output hash of derivations to check whether they need to be built, but the _full_ store path, which notably also includes the _derivation name_. Thus, by ensuring that the derivation name changes if (and only if) relevant inputs change, we can force Nix to rebuild a derivation even if the output hash stays the same, therefore fixing above problem.
+
+This RFC therefore proposes to use a hash of all relevant derivation inputs as the default name of fixed-output derivations. The hash should be computed using `sha256`, and encoded using the [url-safe base64](https://tools.ietf.org/html/rfc4648#section-5), without any leading `=`, leading to a character count of 43. For reference, this hash can be computed in `nix-shell -p openssl` with
+```
+[nix-shell:~]$ echo -n 'example string' | openssl dgst -sha256 -binary | openssl base64 -A | cut -b1-42 | tr +/ -_
+rt-5KzBTohoRT08wGgKjxq1d_1BNEk3CzuYRdiPuxw
+```
```

Comment:
For most sources, the `pname` and `version` attributes are probably enough to invalidate the inputs. One of the benefit is that those could be included in the name, making the store path more readable.

Another benefit of using `pname` and `version` is that it allow to invert the control. Historically the `pname` is first set on the package derivation, and then inherited in the source. In cases where we have a package generator function, like with most languages, the pname could be set on the source, and then inherited by the generator function.

Reply:
I think this rfc is mainly for minority, such as git tarball with `revision` or patch with patchutils arguments, with no `pname` nor `version`. As for normal sources, new `pname` and `version` will generate a new `name`, which could already trigger refetch now. (`Store::makeFixedOutputPath` takes `name` as paramater and generate a different derivation path.)

We may want something other than `hash` or `name` could affect fixedOutputPath.

Reply:
@holymonson if I'm interpreting zimbatm's comment right, he's suggesting that name/pname/version be moved into the source rather than the derivation that builds the source.